### PR TITLE
WIP: Add gpexpand status check in some utility

### PR DESCRIPTION
**Pull request by kainwen**
The following four utilities do not work when we
are in gpexpand phase 1:
    
* gppkg
* gpconfig
* gpcheck
* gpcheckat

Add check for them so that if cluster is expanding
in phase1, they will print error message and exit.

This PR contains a commit picked from PR https://github.com/greenplum-db/gpdb/pull/6554
Please only review 75ba1a09a87e8c37a37a8a08213a14e2e6f2c334
and merge PR 6554 first. Then I will rebase this PR to consider merging.

### CODE REVIEW

File: gpMgmt/bin/gpcheck [hunk 1]
```diff
@@ -849,6 +850,15 @@ def checkUser():
 logger = get_default_logger()
 setup_tool_logging(EXECNAME,getLocalHostname(),getUserName())
 
+# gpcheck should first check gpexpand status
+gpexpand_status = GpexpandRunningStatus()
+if gpexpand_status == "phase1":
+    logger.error("Usage of %s is not supported while the cluster is in a reconfiguration state." % "gpcheck")
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpcheck]: The "reconfiguration state" is not defined in the readme, wiki or the code, please use a proper phase name or a more descriptive name.

Looks like this check is duplicated in the utilities, we should prevent using the hard coded phase name strings.  Please define something like `gppylib.utils.GPEXPAND_PHASE1` for the phase names.

I'm even thinking about moving these checks into a function, `CheckGpexpandStatus(componentName, loggerPrint, exitCode=127, exitOnPhase1=True, exitOnPhase2=False)`.

File: gpMgmt/bin/gpcheckcat [hunk 2]
```diff
@@ -4279,7 +4280,18 @@ def truncate_batch_size(primaries):
 
 
 def main():
+    # gpcheckcat should first check gpexpand status
+    logger.info("hahahaha")
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpcheckcat]: have a nice day :)

File: gpMgmt/bin/gpcheckcat [hunk 3]
```diff
@@ -4279,7 +4280,18 @@ def truncate_batch_size(primaries):
...
+    gpexpand_status = GpexpandRunningStatus()
+    if gpexpand_status == "phase1":
+        myprint("Error: Usage of %s is not supported while the cluster is in a reconfiguration state." % "gpcheckcat") 
+        sys.exit(127)
+    elif gpexpand_status == "unknown gpexpand status":
+        myprint("Error: Unknown gpexpand status, stop %s" % "gpcheckcat")
+        sys.exit(127)
+
     parseCommandLine()
+
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpcheckcat]: remove this line.

File: gpMgmt/bin/gpconfig [hunk 4]
```diff
@@ -500,7 +501,17 @@ def do_show(options):
...
+    gpexpand_status = GpexpandRunningStatus()
+    if gpexpand_status == "phase1":
+        logger.error("Usage of %s is not supported while the cluster is in a reconfiguration state." % "gpcheckcat")
+        sys.exit(127)
+    elif gpexpand_status == "unknown gpexpand status":
+        logger.error("Unknown gpexpand status, stop %s" % "gpcheckcat")
+        sys.exit(127)
+
     options = parseargs()
+
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpconfig]: remove this.

File: src/test/regress/greenplum_schedule [hunk 5]
```diff
@@ -239,7 +239,9 @@ test: oid_wraparound
 test: fts_recovery_in_progress
 test: autovacuum-template0
 
-# Online expand introduce the partial tables, check them if they can run correctly
+# gpexpand introduce the partial tables, check them if they can run correctly
 test: gangsize
+# some utilities do not work while doing gpexpand, check them can print correct message
+test: run_utility_gpexpand_phase1
```
Comment (pivotal-ning-yu) [inline src/test/regress/greenplum_schedule]: Can we run this test in parallel with other tests?

File: src/test/regress/sql/run_utility_gpexpand_phase1.sql [hunk 6]
```diff
@@ -0,0 +1,11 @@
+-- start_matchsubs
+-- m/^.*-/
+-- s/^.*-//
+-- end_matchsubs
+
+\! echo 'EXPANSION_PREPARE_STARTED:<path> to inputfile' > $MASTER_DATA_DIRECTORY/gpexpand.status
+\! gpcheckcat
+\! gpcheck -h `hostname`
+\! gpconfig -c gp_resource_manager -v group
```
Comment (pivotal-ning-yu) [inline src/test/regress/sql/run_utility_gpexpand_phase1.sql]: although we expect gpconfig to fail, but in case we didn't please test against an operation with less side effect, such as `gpconfig -r gp_debug_linger`.

File: gpMgmt/bin/gpcheck (hunk 1 above)
Comment (kainwen) [reply gpMgmt/bin/gpcheck]: This word is proposed in the tracker. @gaos1 What error message is preffered?

File: gpMgmt/bin/gpcheckcat (hunk 2 above)
Comment (kainwen) [reply gpMgmt/bin/gpcheckcat]: Sorry for that. It is debug code. Will remove it soon.

File: src/test/regress/sql/run_utility_gpexpand_phase1.sql (hunk 6 above)
Comment (kainwen) [reply src/test/regress/sql/run_utility_gpexpand_phase1.sql]: Ok.

File: gpMgmt/bin/gpcheck (hunk 1 above)
Comment (kainwen) [reply gpMgmt/bin/gpcheck]: Previously, it is totally encapsulated, but gpcheckcat's logger does not work. Will try to improve it again.

File: src/test/regress/greenplum_schedule (hunk 5 above)
Comment (kainwen) [reply src/test/regress/greenplum_schedule]: In practice, it can be moved into any parallel test groups.
But in logic, it seems this test share noting in common with other test groups.
And this test takes little time. Maybe it is OK to leave it alone?
@pivotal-ning-yu

File: gpMgmt/bin/gpcheck (hunk 1 above)
Comment (kainwen) [reply gpMgmt/bin/gpcheck]: @pivotal-ning-yu Improve the code.

File: src/test/regress/greenplum_schedule (hunk 5 above)
Comment (pivotal-ning-yu) [reply src/test/regress/greenplum_schedule]: In fact there is no need to group tests by function.  But anyway it's ok to keep it in a separate group.

File: gpMgmt/bin/gppylib/utils.py [hunk 7]
```diff
@@ -701,3 +702,31 @@ def escapeDoubleQuoteInSQLString(string, forceDoubleQuote=True):
...
+    elif code >= 200:
+        current_phase = 2
+    else:
+        sys.exit("cannot get gpexpand status, stop %s" % utility)
+
+    if current_phase == 1 and exitOnPhase1:
+        err_msg = ("ERROR: Usage of {utility} is not supported while the"
+                   "cluster is in a reconfiguration state, "
+                   "exit {utility}")
+        sys.exit(err_msg.format(utility=utility))
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gppylib/utils.py]: Is `sys.exit()` messages written to `stderr`?  Is this always suitable for the utilities.  As an alternative we could pass the logger/print function as argument, or raise the error via assert.  My concern on writing to `stderr` is that the messages are not saved in log files.

File: gpMgmt/bin/gpcheckcat [hunk 8]
```diff
@@ -4279,6 +4279,10 @@ def truncate_batch_size(primaries):
 
 
 def main():
+    # gpcheckcat should first check gpexpand status
+    checkGpexpandNotRunning("gpcheckcat", exitOnPhase1=True, exitOnPhase2=False)
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpcheckcat]: I think the `checkGpexpandNotRunning` check should be AFTER the argument  checks.  The user should know whether the arguments are validated before any attempted action by the program.  And interlocking with gpexpand is an action...I'd put the gpexpand running check at L4291, after the version check.

File: gpMgmt/bin/gpcheckcat [hunk 9]
```diff
@@ -46,7 +47,6 @@ oidmap = {}
 # -------------------------------------------------------------------------------
 EXECNAME = os.path.split(__file__)[-1]
 setup_tool_logging(EXECNAME, getLocalHostname(), getUserName())
-very_quiet_stdout_logging()
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpcheckcat]: I'd like to see changes to the log level only around the new `checkGpexpandNotRunning` function, if at all; otherwise, the intent becomes harder to understand.  I think the better design is to have `checkGpexpandNotRunning` return True/False and an error string, and allow the caller of that function to decide the action to take.  It's hard to follow control flow when called functions can exit your program.

File: gpMgmt/bin/gpconfig [hunk 10]
```diff
@@ -500,6 +501,9 @@ def do_show(options):
 
 
 def do_main():
+    # gpconfig should first check gpexpand status
+    checkGpexpandNotRunning("gpconfig", exitOnPhase1=True, exitOnPhase2=False)
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gpconfig]: I think the checkGpexpandNotRunning check should be AFTER the parseargs() check. The user should know whether the arguments are validated before any attempted action by the program. And interlocking with gpexpand is an action...

File: gpMgmt/bin/gppylib/utils.py [hunk 11]
```diff
@@ -701,3 +702,33 @@ def escapeDoubleQuoteInSQLString(string, forceDoubleQuote=True):
     if forceDoubleQuote:
         string = '"' + string + '"'
     return string
+
```
Comment (dkrieger-pvtl) [inline gpMgmt/bin/gppylib/utils.py]: I think the better design is to have checkGpexpandNotRunning return True/False and an error string, and allow the caller of that function to decide the action to take. It's hard to follow control flow when called functions can exit your program.  This can be a discussion: we want uniform behavior so having the check function fail makes sense but then we take away all control/cleanup ability of the caller.

File: gpMgmt/bin/gpcheckcat (hunk 9 above)
Comment (kainwen) [reply gpMgmt/bin/gpcheckcat]: Hi @dkrieger-pvtl Thanks for your review.

I move this function invoke right after gpexpand status checking, which is at the beginning of the main().

File: gpMgmt/bin/gppylib/utils.py (hunk 11 above)
Comment (kainwen) [reply gpMgmt/bin/gppylib/utils.py]: Thanks. Will improve this code.

File: gpMgmt/bin/gpcheckcat [hunk 12]
```diff
@@ -4289,6 +4289,16 @@ def main():
...
 
+    # gpcheckcat should check gpexpand runnning status
+    check_result, msg = checkGpexpandNotRunning("gpcheckcat",
+                                                exitOnPhase1=True,
+                                                exitOnPhase2=False)
+    if not check_result:
+        logger.error(msg)
+        sys.exit(127)
+
+    very_quiet_stdout_logging()
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpcheckcat]: Instead of moving `very_quiet_stdout_logging()` here, we could call `quiet_stdout_logging()` or `enable_verbose_logging()` before `logger.error()` to show the log.

File: gpMgmt/bin/gpconfig [hunk 13]
```diff
@@ -508,6 +509,15 @@ def do_main():
         do_show(options)
 
     elif options.remove or options.change:
+        # gpconfig should check gpexpand runnning status when
+        # it wants to modify configurations
+        check_result, msg = checkGpexpandNotRunning("gpconfig",
+                                                    exitOnPhase1=True,
```
Comment (pivotal-ning-yu) [inline gpMgmt/bin/gpconfig]: As we no longer exit directly inside `checkGpexpandNotRunning()`, I think we need to reconsider the argument names `exitOnPhase1` and `exitOnPhase2`.

Comment (kainwen) [reply gpMgmt/bin/gpconfig]: Sure. Thanks. Will improve the code.

**Comment by kainwen** [comment]
Pushed to feature branch. Thanks.
